// dynamodb-onetable/dist/cjs/index.js
"use strict";
/*
    dynamodb-onetable - DynamoDB wrapper for single table patterns
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.UUID = exports.ULID = exports.UID = exports.Table = exports.OneTableError = exports.OneTableArgError = exports.Model = exports.Expression = void 0;
const Expression_js_1 = require("./Expression.js");
Object.defineProperty(exports, "Expression", { enumerable: true, get: function () { return Expression_js_1.Expression; } });
const Model_js_1 = require("./Model.js");
Object.defineProperty(exports, "Model", { enumerable: true, get: function () { return Model_js_1.Model; } });
const Table_js_1 = require("./Table.js");
Object.defineProperty(exports, "Table", { enumerable: true, get: function () { return Table_js_1.Table; } });
const Error_js_1 = require("./Error.js");
Object.defineProperty(exports, "OneTableError", { enumerable: true, get: function () { return Error_js_1.OneTableError; } });
Object.defineProperty(exports, "OneTableArgError", { enumerable: true, get: function () { return Error_js_1.OneTableArgError; } });
const UID_js_1 = require("./UID.js");
Object.defineProperty(exports, "UID", { enumerable: true, get: function () { return UID_js_1.UID; } });
Object.defineProperty(exports, "ULID", { enumerable: true, get: function () { return UID_js_1.ULID; } });
Object.defineProperty(exports, "UUID", { enumerable: true, get: function () { return UID_js_1.UUID; } });
